Improve input validation for NJT pointwise ops

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138602

Before this PR, NJT would dispatch e.g. `NJT * nested_int` to `mul.Tensor`, wrongly interpreting the SymInt as a tensor and outputting garbage. This PR verifies that there are no nested ints in the list of args before dispatching for pointwise ops.

I originally tried checking that `the number of passed tensor args == the number of func schema tensor args`, but this wrongly disallows `nt * 2`, which (non-intuitively to me at least at first) dispatches via the `mul.Tensor` overload.

jbschlosser committed:
Improve input validation for NJT pointwise ops

[ghstack-poisoned]

davidberard98 reviewed (commented):
will this also reject `nt * 2` or will that pass?

jbschlosser commented:
> will this also reject nt * 2 or will that pass?

good point, it does reject it. I thought we'd hit the `mul.Scalar` overload but no luck

jbschlosser committed:
Update on "Improve input validation for NJT pointwise ops"


Before this PR, we were using the number of passed tensor args to determine whether to dispatch to unary / binary pointwise impls. This had the unintended consequence of allowing e.g. `NJT * nested_int`, wrongly interpreting the SymInt as a tensor and outputting garbage. We should check the func schema itself to be more accurate.

[ghstack-poisoned]

jbschlosser commented:
Updated PR to do a dumber check for any nested ints before dispatching on pointwise ops.

jbschlosser commented:
@pytorchbot merge
